subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/validation/MavenPublicationErrorChecker.java

```diff
@@ -43,14 +48,56 @@ public abstract class MavenPublicationErrorChecker extends PublicationErrorCheck
      * @throws PublishException if the artifacts are modified
      */
     public static void checkThatArtifactIsPublishedUnmodified(PublishArtifact source, DefaultMavenArtifactSet mainArtifacts) {
+        // Do we want to allow _additional_ artifacts to published? If not, we need to completely associate the source and artifacts
+        // at a higher level. As-is, this just verifies that none are removed from the components artifacts.
+        MavenArtifact bestMatch = null;
+        Set<Difference> bestMatchDifferences = null;
         for (MavenArtifact mavenArtifact : mainArtifacts) {
-            if (source.getFile().equals(mavenArtifact.getFile())
-                && source.getExtension().equals(mavenArtifact.getExtension())
-                && Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
+            EnumSet<Difference> difference = EnumSet.noneOf(Difference.class);
+            if (!source.getFile().equals(mavenArtifact.getFile())) {
+                difference.add(Difference.FILE);
+            }
+            if (!source.getExtension().equals(mavenArtifact.getExtension())) {
+                difference.add(Difference.EXTENSION);
+            }
+            if (!Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
```

Comment:
💅 `Objects.equals()` would be better here, to avoid seeing the `nullToEmpty()` checks.

Reply:
Unfortunately, it appears that we map an empty string to `null` in the artifact somewhere, as this causes tests to fail.